### ISSUE TITLE
broken test org.takes.rq.RqMultipartTest.consumesHttpRequest

### ISSUE THREAD

**Issue by dmzaytsev** [yegor256/takes#253]
We have the broken test `org.takes.rq.RqMultipartTest.consumesHttpRequest` in the master branch

>Running org.takes.rq.RqMultipartTest Tests run: 10, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.132 sec <<< FAILURE! - in org.takes.rq.RqMultipartTest consumesHttpRequest(org.takes.rq.RqMultipartTest) Time elapsed: 1.033 sec <<< ERROR! java.io.IOException: Failed POST request to http://localhost:34885/ at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:124) at com.jcabi.http.request.BaseRequest.fetchResponse(BaseRequest.java:299) at com.jcabi.http.request.BaseRequest.fetch(BaseRequest.java:214) at org.takes.rq.RqMultipartTest$2.exec(RqMultipartTest.java:308) at org.takes.http.FtRemote.exec(FtRemote.java:118) at org.takes.rq.RqMultipartTest.consumesHttpRequest(RqMultipartTest.java:303) at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57) at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) at java.lang.reflect.Method.invoke(Method.java:606) at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50) at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12) at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47) at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17) at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78) at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57) at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290) at org.apache.maven.surefire.junitcore.pc.Scheduler$1.run(Scheduler.java:318) at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471) at java.util.concurrent.FutureTask.run(FutureTask.java:262) at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) at java.lang.Thread.run(Thread.java:745) Caused by: java.net.SocketException: Connection reset at java.net.SocketInputStream.read(SocketInputStream.java:196) at java.net.SocketInputStream.read(SocketInputStream.java:122) at java.io.BufferedInputStream.fill(BufferedInputStream.java:235) at java.io.BufferedInputStream.read1(BufferedInputStream.java:275) at java.io.BufferedInputStream.read(BufferedInputStream.java:334) at sun.net.www.MeteredStream.read(MeteredStream.java:134) at java.io.FilterInputStream.read(FilterInputStream.java:133) at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3066) at sun.net.www.protocol.http.HttpURLConnection$HttpInputStream.read(HttpURLConnection.java:3060) at com.jcabi.http.request.JdkRequest$1.body(JdkRequest.java:189) at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:116)

**Comment by m4solutions** [comment]
@dmzaytsev I see the same thing in the merge of my pull request, a question @yegor256  this must be corrected before I can close my already merged pull request? or they are deal with as different issues?

**Comment by super132** [comment]
@dmzaytsev I can see that the error is caused by SocketException that Connection reset. Is there something wrong with the container that rultor runs in? I see no problem running the same tests in Appveyor.

**Comment by dmzaytsev** [comment]
@m4solutions This error is caused by changes in the master branch.
Just wait while the issue  will be resolved

@super132 you see Appveyor's log before merge.
If you merge with master and run `mvn clean install` you will see the error locally

**Comment by super132** [comment]
@dmzaytsev , is the build from the Appvegor also a log before merge? It shows that it is building from a commit in master branch:

https://ci.appveyor.com/project/yegor256/takes/branch/master

**Comment by yegor256** [comment]
@dmzaytsev I can't reproduce this problem locally :( running tests multiple times, still no error

**Comment by yegor256** [comment]
@dmzaytsev can you reproduce it locally?

**Comment by dmzaytsev** [comment]
@yegor256 sure
I reproduced it locally
windows 7 64 bit
5 мая 2015 г. 9:18 AM пользователь "Yegor Bugayenko" <
notifications@github.com> написал:

**Comment by dmzaytsev** [comment]
@yegor256 I can't reproduce the issue with another windows desktop.
seems the error depends on the jdk/os version.
I will get access to that desktop  a few hours later

**Comment by ikhvostenkov** [comment]
@yegor256 @dmzaytsev I have this error randomly on linux64 win java 1.7

**Comment by dmzaytsev** [comment]
@yegor256 what version of JDK uses Rultor ?

**Comment by yegor256** [comment]
@dmzaytsev it's 1.7. you can check it yourself, just install Docker locally and run:

```
docker run -i -t --rm yegor256/rultor /bin/bash
```

You will get into shell with exactly the same environment as we have in rultor builds.

**Comment by dmzaytsev** [comment]
@yegor256 the error is stable on my enviroment 
```
java version "1.8.0_45"
Java(TM) SE Runtime Environment (build 1.8.0_45-b15)
Java HotSpot(TM) 64-Bit Server VM (build 25.45-b02, mixed mode)
Windows 7 64bit SP 1
```
I see two issues:
*  the regular expression ` ".*[^a-z]name=\"([^\"]+)\".*"` (RqMultipart.java:109) doesn't extract `f-1` from the string `"Content-Disposition: form-data; name=\"f-1\""`(RqMultipartTest.java:318)
* if the name is not found so we have broken connection in FtRemote instead exception. I think it is a bug. You can easy reproduce it that way
```
public void consumesHttpRequest() throws IOException {
        final Take take = new Take() {
            @Override
            public Response act(final Request req) throws IOException {
                return new RsText(
                    new RqPrint(
                        new RqMultipart.Smart(
                            new RqMultipart.Base(req)
                        ).single("f2")
                    ).printBody()
                );
            }
        };
        new FtRemote(take).exec(
            // @checkstyle AnonInnerLengthCheck (50 lines)
            new FtRemote.Script() {
                @Override
                public void exec(final URI home) throws IOException {
                    new JdkRequest(home)
                        .method("POST")
                        .header(
                            "Content-Type",
                            "multipart/form-data; boundary=AaB0zz"
                        )
                        .body()
                        .set(
                            Joiner.on(RqMultipartTest.CRLF).join(
                                "--AaB0zz",
                                "Content-Disposition: form-data; name=\"f1\"",
                                "",
                                "my picture",
                                "--AaB0zz--"
                            )
                        )
```

**Comment by yegor256** [comment]
@dmzaytsev many thanks!

**Comment by yegor256** [comment]
@davvd this is urgent

**Comment by super132** [comment]
@yegor256 , just an update to this issue. I found that a PR #113 has been merged by Rultor successfully. Please check the log below:

http://www.rultor.com/t/3570-99185933

**Comment by davvd** [comment]
@dmzaytsev I am aware of the task, give me some time to find a developer...

**Comment by davvd** [comment]
@dmzaytsev I added milestone `1.0` to this issue, let me know if there has to be something else

**Comment by davvd** [comment]
> @davvd this is urgent

@yegor256 sure, thanks, I added "urgent" tag to it

**Comment by davvd** [comment]
@bertram25 This task is yours, please go ahead keeping in mind [this](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html). If any questions, don't hesitate to ask right here. **30 mins** is the budget of the task. This is exactly how much will be paid when the problem is [solved](http://www.xdsd.org/2014/04/15/definition-of-done.html) (no matter how much time you will actually spend). See [this](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html) for more information

**Comment by davvd** [comment]
@dmzaytsev thanks for reporting! I topped your account for 15 mins, transaction 56895756

**Comment by Bertram25** [comment]
@dmzaytsev @yegor256  Hi,

> the regular expression ".*[^a-z]name=\"([^\"]+)\".*" (RqMultipart.java:109) doesn't extract f-1 from the string "Content-Disposition: form-data; name=\"f-1\""(RqMultipartTest.java:318)

For the record, I confirm the current code can't get the `f-1` name and fails systematically.
Simply using `matcher.find()` instead of `matcher.matches()` is fixing the issue for me btw. (In 
https://github.com/yegor256/takes/blob/master/src/main/java/org/takes/rq/RqMultipart.java#L311 )
As it seems finding the next matching group, instead of trying to search the whole string does give proper result.

But with this fix, the test still fails at random with a socket exception. (Using OpenJDK 1.7 / Linux 64bit) Thus the bug is IMHO something else.

> if the name is not found so we have broken connection in FtRemote instead exception. I think it is a bug. You can easy reproduce it that way

While I agree about the existence of that bug, I do think it's a separate issue.
I also so think FtRemote should be fixed first in that regards to permit better understanding of this issue. Do you agree about it?

Best regards,

**Comment by yegor256** [comment]
@Bertram25 if you see some bug in the existing code, don't hesitate to report it in a new ticket. You don't need anyone's permission for that.

**Comment by Bertram25** [comment]
@yegor256 Thanks, just did. (Still getting used to it.) 

Btw, I know I'm "cutting every corners" here and I hope it's fine that way.

IMHO, the bug #291 needs to be fixed *first* but won't suffice to fix this one, hence why it isn't a duplicate.
If you think my reasoning is correct, then I'll stop working on the current bug (as the name parameter value issue part is solved), and I'll wait for #291 to be solved first.

**Comment by Bertram25** [comment]
@davvd Hi David,

Since I've got no answer from @yegor256 yet, I just wanted to say that I've stopped working on this issue because #291 has to be solved first IMO. I'll start again on this one, once it is solved unless otherwise told.

Regards,

**Comment by yegor256** [comment]
@davvd we're waiting for #291 here

**Comment by davvd** [comment]
> @davvd we're waiting for #291 here

@yegor256 yes, waiting for #291

**Comment by davvd** [comment]
@louzar This task is yours, please go ahead keeping in mind [this](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html). If any questions, don't hesitate to ask right here; Budget here is **30 mins** (keep [this principle](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html) in mind)

**Comment by louzar** [comment]
@davvd  I just get notified 2 days ago that I'm on this project, does this issue still for me

**Comment by davvd** [comment]
@komaz please go ahead with this task, it's yours

**Comment by komaz** [comment]
@davvd waiting for #291

**Comment by davvd** [comment]
> @davvd waiting for #291 

@komaz OK, let's wait for #291

**Comment by komaz** [comment]
@dmzaytsev is it ok that this issue is not unassigned?

**Comment by dmzaytsev** [comment]
@komaz I see the `@komaz` tag here
It means you assigned to this ticket

**Comment by komaz** [comment]
@dmzaytsev yeah, but this issue is blocked by #291, I thought such issues should be unassigned to avoid penalty, e.g. https://github.com/yegor256/netbout/issues/602

**Comment by dmzaytsev** [comment]
@komaz yegor256/netbout#602 is postponed

but @davvd wrote to you above
>@komaz OK, let's wait for #291

You shouldn't be fined, I hope :)

Anyway @yegor256 and @davvd to manage this project
Ask them if you not sure

**Comment by yegor256** [comment]
@komaz yes, everything is fine here, we're blocked by #291

**Comment by Bertram25** [comment]
@komaz FYI, issue #291 is now closed.
Best regards,

**Comment by davvd** [comment]
@dmzaytsev I think we can continue, impediment #291 is closed

**Comment by dmzaytsev** [comment]
@komaz FYI

> @dmzaytsev I think we can continue, impediment #291 is closed

**Comment by komaz** [comment]
@dmzaytsev fixed in #329

**Comment by dmzaytsev** [comment]
@komaz thanks!

**Comment by davvd** [comment]
@ypshenychka please, let us know what do you think about this ticket, according to our [QA rules](http://www.teamed.io/qa.html)

**Comment by ypshenychka** [comment]
@davvd Quality is good.

**Comment by davvd** [comment]
> @davvd Quality is good.

@ypshenychka thanks for the review and a good remark

**Comment by davvd** [comment]
@komaz just added 10 mins to @ypshenychka (for QA), payment ID is `58676446`; Much obliged! I have added **38 mins** to your account in payment "AP-5MH89111A09400627", 57 hours and 30 mins spent; the bonus for fast delivery (m=3450); added +38 to your rating, now it is equal to [+458](http://www.netbout.com/b/35322?open=rating)
